server/src/main/java/org/opensearch/index/remote/RemoteIndexPathUploader.java

```diff
@@ -0,0 +1,261 @@
+            } catch (InterruptedException exception) {
+                exceptionList.forEach(exception::addSuppressed);
+                RemoteStateTransferException ex = new RemoteStateTransferException(
+                    String.format(Locale.ROOT, TIMEOUT_EXCEPTION_MSG, indexNames),
+                    exception
+                );
+                actionListener.onFailure(ex);
+                return;
+            }
+            if (exceptionList.size() > 0) {
+                RemoteStateTransferException ex = new RemoteStateTransferException(
+                    String.format(Locale.ROOT, UPLOAD_EXCEPTION_MSG, indexNames)
+                );
+                exceptionList.forEach(ex::addSuppressed);
+                actionListener.onFailure(ex);
+                return;
+            }
+            success = true;
+            actionListener.onResponse(null);
+        } finally {
+            long tookTimeNs = System.nanoTime() - startTime;
+            logger.trace(new ParameterizedMessage("executed beforeNewIndexUpload status={} tookTimeNs={}", success, tookTimeNs));
+        }
+
+    }
+
+    private void writeIndexPathAsync(IndexMetadata idxMD, CountDownLatch latch, List<Exception> exceptionList) throws IOException {
+        if (isTranslogSegmentRepoSame) {
+            // If the repositories are same, then we need to upload a single file containing paths for both translog and segments.
+            writePathToRemoteStore(idxMD, translogRepository, latch, exceptionList, COMBINED_PATH);
+        } else {
+            // If the repositories are different, then we need to upload one file per segment and translog containing their individual
+            // paths.
+            writePathToRemoteStore(idxMD, translogRepository, latch, exceptionList, TRANSLOG_PATH);
+            writePathToRemoteStore(idxMD, segmentRepository, latch, exceptionList, SEGMENT_PATH);
+        }
+    }
+
+    private void writePathToRemoteStore(
+        IndexMetadata idxMD,
+        BlobStoreRepository repository,
+        CountDownLatch latch,
+        List<Exception> exceptionList,
+        Map<RemoteStoreEnums.DataCategory, List<RemoteStoreEnums.DataType>> pathCreationMap
+    ) throws IOException {
+        Map<String, String> remoteCustomData = idxMD.getCustomData(IndexMetadata.REMOTE_STORE_CUSTOM_KEY);
+        RemoteStoreEnums.PathType pathType = RemoteStoreEnums.PathType.valueOf(remoteCustomData.get(RemoteStoreEnums.PathType.NAME));
+        RemoteStoreEnums.PathHashAlgorithm hashAlgorithm = RemoteStoreEnums.PathHashAlgorithm.valueOf(
+            remoteCustomData.get(RemoteStoreEnums.PathHashAlgorithm.NAME)
+        );
+        String indexUUID = idxMD.getIndexUUID();
+        int shardCount = idxMD.getNumberOfShards();
+        BlobPath basePath = repository.basePath();
+        BlobContainer blobContainer = repository.blobStore().blobContainer(basePath.add(RemoteIndexPath.DIR));
+        REMOTE_INDEX_PATH_FORMAT.writeAsyncWithUrgentPriority(
```

Comment:
Do we need any other repository details? What is the expcetation from caller while reading these files? What information should caller have?

Reply:
> Do we need any other repository details?

The repository is supposed to be the one storing anything related to translog or segment data/metadata/lock_files.

> What is the expcetation from caller while reading these files?

We have not implemented the read yet in OS. The file is supposed to be used externally for now.

> What information should caller have?

Not sure if I understood you completely.

Reply:
@sachinpkale as discussed, we are uploading the file in the same repository where the path also exists.

Reply:
Resolving this, please re-open for any follow up questions/feedback.